File: javascript/bike.js
Util = {
  RADIANS: 180/3.14169,

  kilometersBetween: function(lat1, lng1, lat2, lng2) {
    var a1 = lat1 / Util.RADIANS,
        a2 = lng1 / Util.RADIANS,
        b1 = lat2 / Util.RADIANS,
        b2 = lng2 / Util.RADIANS,
        t1 = Math.cos(a1) * Math.cos(a2) * Math.cos(b1) * Math.cos(b2),
        t2 = Math.cos(a1) * Math.sin(a2) * Math.cos(b1) * Math.sin(b2),
        t3 = Math.sin(a1) * Math.sin(b1)

    return 6366 * Math.acos(t1 + t2 + t3)
  }
}

City = function(name, lat, lng) {
  this.name = name
  this.lat  = lat
  this.lng  = lng
}
City.prototype.kilometersTo = function(other) {
  return Util.kilometersBetween(
    this.lat,  this.lng,
    other.lat, other.lng
  )
}
City.prototype.roads = function() {
  if (this._roads) return this._roads
  var roads = [],
      self = this
  Road.All.forEach(function(road) {
    if (road.a == self || road.b == self)
      roads.push(road)
  })
  this._roads = roads
  return roads
}
City.prototype.adjacentCities = function() {
  if (this._cities) return this._cities
  var cities = [],
      roads = this.roads()
  for (var i in roads)
    cities.push(roads[i].theCityOpposite(this))
  this._cities = cities
  return cities
}
City.prototype.toString = function() {
  return this.name
}

Berlin    = new City('Berlin',    52.482668, 13.359275)
Paris     = new City('Paris',     48.980405, 2.2851849)
Milan     = new City('Milan',     45.520543, 9.1419459)
Frankfurt = new City('Frankfurt', 50.078848, 8.6349115)
Munich    = new City('Munich',    48.166229, 11.558089)
Zurich    = new City('Zurich',    47.383444, 8.5254142)
Tours     = new City('Tours',     47.413572, 0.6810506)
Lyon      = new City('Lyon',      45.767122, 4.8339568)
Vienna    = new City('Vienna',    48.224431, 16.389240)
Prague    = new City('Prague',    50.092396, 14.436144)
Krakow    = new City('Krakow',    50.050363, 19.928578)
Warsaw    = new City('Warsaw',    52.254756, 21.005968)
Hamburg   = new City('Hamburg',   53.539699, 9.9977143)
Antwerp   = new City('Antwerp',   51.220613, 4.3954882)
Torino    = new City('Torino',    45.105321, 7.6451957)
Rome      = new City('Rome',      42.032845, 12.390408)

Road = function(a, b) {
  this.a = a
  this.b = b
  this.distance = a.kilometersTo(b)
}
Road.prototype.theCityOpposite = function(city) {
  return city == this.a ? this.b : this.a
}
Road.prototype.toString = function() {
  return this.a.toString() + " | "
       + this.b.toString() + " ("
       + this.distance.toFixed(3) + " km)"
}
Road.All = []
Road.All.contains = function(x) {
  // mimicking a Set here
  return this.hasOwnProperty(x.toString())
}
Road.All.add = function(element) {
  if (!this.contains(element)) {
    this.push(element)
    this[element.toString()] = true
  }
}
Road.between = function(a, b) {
  for (var i in Road.All) {
    var r = Road.All[i]
    console.log("checking: ", r)
    console.log(r.a.name, r.b.name)
    var road = Road.All[i]
    if (   (road.a == this.a && road.b == this.b)
        || (road.a == this.b && road.b == this.a)) {
      console.log(i, "returning road: ", road)
      return road 
    }

  }
  return "No road between " + a + " and " + b + "!"
}

Road.All.add(new Road(Hamburg,   Berlin))
Road.All.add(new Road(Hamburg,   Antwerp))
Road.All.add(new Road(Hamburg,   Frankfurt))
Road.All.add(new Road(Berlin,    Warsaw))
Road.All.add(new Road(Berlin,    Prague))
Road.All.add(new Road(Antwerp,   Paris))
Road.All.add(new Road(Paris,     Tours))
Road.All.add(new Road(Paris,     Lyon))
Road.All.add(new Road(Paris,     Zurich))
Road.All.add(new Road(Paris,     Frankfurt))
Road.All.add(new Road(Frankfurt, Prague))
Road.All.add(new Road(Krakow,    Warsaw))
Road.All.add(new Road(Krakow,    Prague))
Road.All.add(new Road(Krakow,    Vienna))
Road.All.add(new Road(Vienna,    Munich))
Road.All.add(new Road(Vienna,    Prague))
Road.All.add(new Road(Zurich,    Milan))
Road.All.add(new Road(Lyon,      Torino))
Road.All.add(new Road(Torino,    Milan))
Road.All.add(new Road(Torino,    Rome))
Road.All.add(new Road(Milan,     Rome))

Start = Rome
End   = Berlin



//result = your_code()
// Your job: make `result` a list of cities from Rome to Berlin

visited = {}
function travel(start) {
  frontier = [[start]]
  while (frontier.length) {
    var path = frontier.shift(),
        city = path[path.length-1]
    console.log(path)
    visited[city] = true
    if (city == End)
      return path
    for (var i in city.adjacentCities()) {
      var newCity = city.adjacentCities()[i],
          newPath = path.slice(0)
      if (visited.hasOwnProperty(newCity)) {
        console.log("skipping ", city)
      } else {
        console.log("pushing " + newCity)
        newPath.push(newCity)
        frontier.push(newPath)
      }
    }
  }
}
result = travel(Start)


totalDistance = 0
for (var i in result) {
  var city = result[i]
  var nextCity = result[parseInt(i)+1]
  console.log(i, result.length, city)
  console.log("nextcity: ", nextCity)

  if (nextCity) {
    var road = Road.between(city, nextCity),
        distance = road.distance
    console.log(road, distance)
    totalDistance += distance
    console.log(" -> " + distance.toFixed(3))
  }
  else {
    console.log('no nextCity')
  }
}
console.log("arrived in " + result.length + " steps ")
console.log("(" + totalDistance.toFixed(3) + " km)")

File: python/bike.py
import math
import random

RADIANS = 180/3.14169
def kilometers_between(lat1, lng1, lat2, lng2):
    a1 = lat1 / RADIANS
    a2 = lng1 / RADIANS
    b1 = lat2 / RADIANS
    b2 = lng2 / RADIANS

    t1 = math.cos(a1) * math.cos(a2) * math.cos(b1) * math.cos(b2)
    t2 = math.cos(a1) * math.sin(a2) * math.cos(b1) * math.sin(b2)
    t3 = math.sin(a1) * math.sin(b1)

    return 6366 * math.acos(t1 + t2 + t3)

class City:
    def __init__(self, name, lat, lng):
        self.name = name
        self.lat  = lat
        self.lng  = lng


    def kilometers_to(self, other):
        return kilometers_between(self.lat, self.lng, other.lat, other.lng)

    def roads(self):
        return filter(lambda road:road.a is self or road.b is self, Roads)

    def adjacent_cities(self):
        return [road.the_city_opposite(self) for road in self.roads()]

    def __repr__(self):
        return self.name


class Road:
    def __init__(self, a, b):
        if a is b:
            raise Exception("Roads must have two distinct cities")
        self.a = a
        self.b = b
        self.distance = a.kilometers_to(b)

    def the_city_opposite(self, city):
        if city is self.a:
            return self.b
        elif city is self.b:
            return self.a
        else:
            raise Exception("%s isn't connected to %s" % (city, self))


    def __repr__(self):
        return "%s - %s (%.0f) km" % (a, b, self.distance)


    @staticmethod
    def between(a, b):
        for road in Roads:
            if (road.a is a and road.b is b) or (road.a is b and road.b is a):
                return road
        raise Exception("There is no road between %s and %s" % (a, b))


Berlin    = City('Berlin',    52.482668, 13.359275)
Paris     = City('Paris',     48.980405, 2.2851849)
Milan     = City('Milan',     45.520543, 9.1419459)
Frankfurt = City('Frankfurt', 50.078848, 8.6349115)
Munich    = City('Munich',    48.166229, 11.558089)
Zurich    = City('Zurich',    47.383444, 8.5254142)
Tours     = City('Tours',     47.413572, 0.6810506)
Lyon      = City('Lyon',      45.767122, 4.8339568)
Vienna    = City('Vienna',    48.224431, 16.389240)
Prague    = City('Prague',    50.092396, 14.436144)
Krakow    = City('Krakow',    50.050363, 19.928578)
Warsaw    = City('Warsaw',    52.254756, 21.005968)
Hamburg   = City('Hamburg',   53.539699, 9.9977143)
Antwerp   = City('Antwerp',   51.220613, 4.3954882)
Torino    = City('Torino',    45.105321, 7.6451957)
Rome      = City('Rome',      42.032845, 12.390408)


Roads = set()
Roads.add(Road(Hamburg,   Berlin))
Roads.add(Road(Hamburg,   Antwerp))
Roads.add(Road(Hamburg,   Frankfurt))
Roads.add(Road(Berlin,    Warsaw))
Roads.add(Road(Berlin,    Prague))
Roads.add(Road(Antwerp,   Paris))
Roads.add(Road(Paris,     Tours))
Roads.add(Road(Paris,     Lyon))
Roads.add(Road(Paris,     Zurich))
Roads.add(Road(Paris,     Frankfurt))
Roads.add(Road(Frankfurt, Prague))
Roads.add(Road(Krakow,    Warsaw))
Roads.add(Road(Krakow,    Prague))
Roads.add(Road(Krakow,    Vienna))
Roads.add(Road(Vienna,    Munich))
Roads.add(Road(Vienna,    Prague))
Roads.add(Road(Zurich,    Milan))
Roads.add(Road(Lyon,      Torino))
Roads.add(Road(Torino,    Milan))
Roads.add(Road(Torino,    Rome))
Roads.add(Road(Milan,     Rome))


Start = Rome
End   = Berlin

result = [] # make a list of cities from Rome to Berlin

total_distance = 0
for idx, city in enumerate(result):
    if idx < len(result) - 1:
        next_city = result[idx+1]
        distance = Road.between(city, next_city).distance
        total_distance += distance
        print "%s - %s (%.0f) km" % (city, next_city, distance)


print "arrived in %d steps (%.0f km)" % (len(result), total_distance)
print ""
print ""

